### ISSUE TITLE
Help setting up scipy for development?

### ISSUE THREAD

**Issue by Hvass-Labs** [scipy/scipy#5893]
Hello,

I'd like to contribute a few things to scipy. I haven't contributed to an open-source project on github before and I'm fairly new to both python and linux. It has been very challenging to get all this working and I've considered several times if it was worth the time and effort. I think you may get more contributors if this was either made easier or explained better.

The scipy / numpy manual has almost no details e.g. regarding virtualenv:

http://docs.scipy.org/doc/numpy-dev/dev/development_environment.html

So I was hoping someone would share a more detailed example of how they have scipy set up for development.

Here's what I've done: First I've forked scipy on github. Then I've cloned that fork to my local PC and placed it in ~/development/scipy/ I will then make a git-branch and add source-code to scipy in that branch. Later I will push the branch to my github fork of scipy and then make a pull-request for merging it into the scipy repository. (A week ago this would have been gibberish to me :-)

But first I want to make a pycharm project in ~/development/scipy_test/ which imports the scipy installation in ~/development/scipy/ instead of the one I've previously installed with pip. I prefer to use pycharm and I'd like to emulate how I would use the source-code I'm adding to scipy in a real-world project, so I can see everything works as expected (I've also made unit-tests but they are of a different nature).

I've tried adding a symbolic link to make python include my development version of scipy:

```shell
ln -s ~/development/scipy/scipy ~/development/scipy_test/scipy
```

But when I import scipy in python3 I get the following error:

```shell
~/development/scipy-test $ python3
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy
Traceback (most recent call last):
  File "/home/magnus/development/scipy-test/scipy/__init__.py", line 98, in <module>
    from scipy.__config__ import show as show_config
ImportError: No module named 'scipy.__config__'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/magnus/development/scipy-test/scipy/__init__.py", line 103, in <module>
    raise ImportError(msg)
ImportError: Error importing scipy: you cannot import scipy while
        being in scipy source directory; please exit the scipy source
        tree first, and relaunch your python intepreter.
>>> 
```

So apparently scipy doesn't like to be imported using a symbolic link.

I've tried creating a virtualenv in pycharm but I cannot get it working because pycharm refuses to use the existing dir ~/development/scipy/ which was created using git clone.

Can someone explain in more detail how scipy is normally set up for development? And how would you set it up with the pycharm test-project that I want to use?

Thanks.

**Comment by nmayorov** [comment]
@Hvass-Labs I suggest you to try managing dev versions using PYTHONPATH. Maybe it's not the most flexible approach, but it works for me.

To get the dev versions of Python packages in your shell, you need to extend PYTHONPATH in your shell config file. For example, I use OSX and I have the following line in `.bash_profile` (AFAIK its Linux equivalent is `.bashrc`) to enable scipy-dev version:
```
export PYTHONPATH="$HOME/PythonDev/scipy/:$PYTHONPATH"
```
Now `python -c "import scipy; print(scipy.__version__)"` gives me `0.18.0.dev0+1fee01c` (and I do have another version of scipy from anaconda, which is now shaded because of PYTHONPATH.)

PyCharm currently doesn't pick your PYTHONPATH from `.bash_profile` (and equivalents), so what I usually do when I need a dev version running in PyCharm, I go to `Run>Edit Configurations>Defaults>Python` and add the required PYTHONPATH to `Environment variables`. There is some kind of a bug(s) in PyCharm PYTHONPATH management which results in this inconvenience, for now I suggest to use this manual method.

Hope it was helpful.

**Comment by andyfaff** [comment]
I use pycharm for scipy development.  Once you've cloned scipy open the root directory in pycharm, which will set it up as a project. I then create a specific conda environment to act as a development environment for scipy. Once you've done that you can select that development environment in the pycharm configuration  preferences to act as the python interpreter.

**Comment by andyfaff** [comment]
That means no messing around with PYTHONPATH,  etc, pycharm does it all for you. Once I've made changes to the codebase I install it into the conda environment and start testing. I'm not sure if it's possible to start testing without the install step.

**Comment by pv** [comment]
Closing, not a bug (feel free to continue discussion though).

**Comment by Hvass-Labs** [comment]
Thanks for the quick replies. Unfortunately I can't get it working. I wonder if I may have done something wrong while installing all these packages and dependencies using pip and apt-get.

If I setup $PYTHONPATH to include the scipy path that I cloned from github, and then run python and import scipy I get the following error:

```bash
~ $ echo $PYTHONPATH
/home/magnus/development/scipy/:
~ $ python2
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/magnus/development/scipy/scipy/__init__.py", line 103, in <module>
    raise ImportError(msg)
ImportError: Error importing scipy: you cannot import scipy while
        being in scipy source directory; please exit the scipy source
        tree first, and relaunch your python intepreter.
>>> 
```

The same happens when I run python 3.4.3 instead.

I've now installed Anaconda v. 2.5.0 which uses python 3.5.1. It's installed with the default options in ~/anaconda3/. When importing scipy in this python version, it shows the same error as above.

If I run sudo python2 or sudo python3 then it uses python 2.7 and 3.4 (i.e. not anaconda) and imports the scipy that I had previously installed using pip and apparently ignores the $PYTHONPATH that I set for using scipy from ~/development/scipy/ That works fine but then it's scipy v 0.17.0 and scipy.__path__ is /usr/local/lib/python2.7/dist-packages/scipy when I run sudo python2 and /usr/local/lib/python3.4/dist-packages/scipy when I run sudo python3.

I can also execute ~/development/scipy/runtests.py using python2 and python3 but not using the anaconda version of python, which gives the following error:

```bash
~/development/scipy $ python runtests.py -v

(... many warnings snipped ...)

gcc -pthread -shared -L/home/magnus/anaconda3/lib -Wl,-rpath=/home/magnus/anaconda3/lib,--no-as-needed build/temp.linux-x86_64-3.5/scipy/cluster/_vq.o -L/home/magnus/anaconda3/lib -L/home/magnus/anaconda3/lib -Lbuild/temp.linux-x86_64-3.5 -lmkl_lapack95_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lpython3.5m -o build/lib.linux-x86_64-3.5/scipy/cluster/_vq.cpython-35m-x86_64-linux-gnu.so
/usr/bin/ld: cannot find -lmkl_lapack95_lp64
collect2: error: ld returned 1 exit status
/home/magnus/anaconda3/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'install_requires'
  warnings.warn(msg)
/home/magnus/anaconda3/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'test_suite'
  warnings.warn(msg)
/home/magnus/anaconda3/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'setup_requires'
  warnings.warn(msg)
/home/magnus/anaconda3/lib/python3.5/site-packages/numpy/distutils/system_info.py:635: UserWarning: Specified path  is invalid.
  warnings.warn('Specified path %s is invalid.' % d)
Running from scipy source directory.
/home/magnus/anaconda3/lib/python3.5/site-packages/numpy/distutils/system_info.py:635: UserWarning: Specified path /usr/local/include/python3.5m is invalid.
  warnings.warn('Specified path %s is invalid.' % d)
/home/magnus/anaconda3/lib/python3.5/site-packages/numpy/distutils/system_info.py:635: UserWarning: Specified path /usr/include/python3.5m is invalid.
  warnings.warn('Specified path %s is invalid.' % d)
/usr/bin/ld: cannot find -lmkl_lapack95_lp64
collect2: error: ld returned 1 exit status
error: Command "gcc -pthread -shared -L/home/magnus/anaconda3/lib -Wl,-rpath=/home/magnus/anaconda3/lib,--no-as-needed build/temp.linux-x86_64-3.5/scipy/cluster/_vq.o -L/home/magnus/anaconda3/lib -L/home/magnus/anaconda3/lib -Lbuild/temp.linux-x86_64-3.5 -lmkl_lapack95_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lpython3.5m -o build/lib.linux-x86_64-3.5/scipy/cluster/_vq.cpython-35m-x86_64-linux-gnu.so" failed with exit status 1

Build failed!
```

Any ideas? Should I install or uninstall something?

Thanks.

**Comment by nmayorov** [comment]
@Hvass-Labs did you actually build scipy-dev before trying to import it? To build in-place run `python setup.py build_ext -i` in the cloned folder. 

From the second log you provided it looks like you will have troubles building it. But that's a different story. At least confirm that you tried to build it.

**Comment by ev-br** [comment]
In addition to what Nikolay and Andrew said, this error

```
ImportError: Error importing scipy: you cannot import scipy while
        being in scipy source directory; please exit the scipy source
        tree first, and relaunch your python intepreter.
```

is just what it says on the box: you cannot import scipy from the source directory. Using the command line:

```
br@duneyrr:~/repos/scipy$ pwd
/home/br/repos/scipy
br@duneyrr:~/repos/scipy$ python
Python 2.7.11 (default, Dec 14 2015, 22:56:59) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "scipy/__init__.py", line 103, in <module>
    raise ImportError(msg)
ImportError: Error importing scipy: you cannot import scipy while
        being in scipy source directory; please exit the scipy source
        tree first, and relaunch your python intepreter.
>>> exit()
br@duneyrr:~/repos/scipy$ 
br@duneyrr:~/repos/scipy$ cd ..
br@duneyrr:~/repos$ python
Python 2.7.11 (default, Dec 14 2015, 22:56:59) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy
>>>   # works!
```

I suppose something similar happens if you use pycharm or whatnot --- try running it from some other directory.

Also, now that you mention sudo and apt-get: you are really better off if you *avoid*, I repeat *avoid*, using sudo for installing scipy. Basically, `sudo pip` is something you simply never type. If you really want to install a python package system-wide, use `pip install --user`. (But really, you almost always want to use a virtualenv or a conda env).

**Comment by Hvass-Labs** [comment]
@nmayorov I haven't tried building it before. When I try I get the following output:

```bash
~/development/scipy $ python setup.py build_ext -i

(... many status messages snipped ...)

gcc -pthread -shared -L/home/magnus/anaconda3/lib -Wl,-rpath=/home/magnus/anaconda3/lib,--no-as-needed build/temp.linux-x86_64-3.5/scipy/cluster/_vq.o -L/home/magnus/anaconda3/lib -L/home/magnus/anaconda3/lib -Lbuild/temp.linux-x86_64-3.5 -lmkl_lapack95_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lpython3.5m -o scipy/cluster/_vq.cpython-35m-x86_64-linux-gnu.so
/usr/bin/ld: cannot find -lmkl_lapack95_lp64
collect2: error: ld returned 1 exit status
/usr/bin/ld: cannot find -lmkl_lapack95_lp64
collect2: error: ld returned 1 exit status
error: Command "gcc -pthread -shared -L/home/magnus/anaconda3/lib -Wl,-rpath=/home/magnus/anaconda3/lib,--no-as-needed build/temp.linux-x86_64-3.5/scipy/cluster/_vq.o -L/home/magnus/anaconda3/lib -L/home/magnus/anaconda3/lib -Lbuild/temp.linux-x86_64-3.5 -lmkl_lapack95_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lpython3.5m -o scipy/cluster/_vq.cpython-35m-x86_64-linux-gnu.so" failed with exit status 1
```

How do I get rid of this error? Do I need to apt-get or pip install something? (And no sudo install this time?)

@ev-br I was not actually inside the ~/development/scipy/ directory when the error occurred. The error occured if I made a symbolic link to the directory e.g. from ~/development/scipy-test/scipy/ to ~/development/scipy/. The error also occured when I set the $PYTHONPATH as suggested above and started python e.g. from my home-folder.

However, I cannot reproduce the error now. Now I can set the $PYTHONPATH to ~/development/scipy/ and I can import that version of scipy in python 2.7, 3.4, and anaconda. It prints the correct version and path. Great!

It appears I have missed the build-step when I cloned the github. I've taken a look at the instructions for scipy / numpy:

http://docs.scipy.org/doc/numpy-dev/user/building.html

Those instructions are really written for people who are already proficient with all these tools and work-flows. I think it may be worth considering if the instructions could be improved for people who are not already experts at these things.

Thanks for the info on sudo pip being bad. I can't quite recall why I did that, probably just because I was trying a bunch of things to get it all working, and that may have worked for some reason. I now read on a forum that sudo pip is bad because it gives root-access to the package's setup script which may contain malicious code. That makes sense.

Follow up questions: Are those packages that I installed with sudo pip now installed in a 'wrong' directory? Should I remove them and re-install just using pip without sudo?

Also, could you elaborate on why you recommend always using virtualenv or conda environments? I'm just one person using one machine. I don't work on client projects that require me to switch between version-numbers of different packages, so I can generally just install and use the latest stable builds of the packages, which can be installed for my personal account or machine-wide and used on all the projects I work on. In case I need to override a package - e.g. in this case where I'd like to override scipy, it appears to be sufficient that I use $PYTHONPATH. So why do I need virtualenv or conda env?

Thanks again for the quick replies and I appreciate everyone's patience. It's a pretty steep learning curve with these tools, at least for me :-)

**Comment by nmayorov** [comment]
@Hvass-Labs to get the full functionality you absolutely need to build all binary files used in scipy.  

The errors tell that the linker tries to use MKL LAPACK library which don't exist on your system. Your best bet to sort it out is to read here http://scipy.org/scipylib/building/linux.html, it shouldn't be difficult on a Linux system. And I'm sure @ev-br or someone else can help you if you will stuck.

**Comment by Hvass-Labs** [comment]
I've looked at the build instructions for numpy / scipy:

http://scipy.org/scipylib/building/linux.html

One of the first requirements is to install BLAS and LAPACK.

These are the "basic steps" of installing BLAS:

http://math-atlas.sourceforge.net/atlas_install/node6.html

This is the "quick" installation guide for LAPACK:

http://www.netlib.org/lapack/lawn81/

These do not appear to be easy to install.

Guys, although I appreciate the help you've given in this thread, it has been very frustrating trying to get scipy working so I can contribute a few things. I think it would be well worth the effort if you either simplify the procedure, or significantly improve the instructions. Currently there are huge barriers to contributing to scipy, which perhaps explains why the community is so small. You cannot assume that experts in various numerical algorithms are also experts at setting up linux build environments, using github, etc. So it might be a wise investment to make it much easier to contribute.

Here's an example of installation instructions that are much easier to follow and understand:

https://www.tensorflow.org/versions/r0.7/get_started/os_setup.html

(Interestingly, they want you to use 'sudo pip' for installing TensorFlow.)

I think I'll take a break from trying to get scipy working before I throw my PC out the window in frustration :-) Perhaps I'll return later and take another stab at it.

**Comment by ev-br** [comment]
These instructions have been reworked recently: https://github.com/scipy/scipy/pull/5830

I would do something along the lines of `sudo apt-get build-dep python-scipy` to bring in the dependencies, then ( in a virtualenv), `pip install numpy nose cython`, `python setup.py build`, `python setup.py install`.

**Comment by Hvass-Labs** [comment]
Thanks for the quick response.

I've tried what you suggest but I get the following error:

```bash
~ $ sudo apt-get build-dep python-scipy
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: You must put some 'source' URIs in your sources.list
```

Googling this error I apparently need to edit /etc/apt/sources.list and add a server so apt-get will download source-code (?) These are the full instructions:

https://help.ubuntu.com/community/Repositories/CommandLine

Apparently I need to add these two lines to the top of my /etc/apt/sources.list:

deb http://ch.archive.ubuntu.com/ubuntu/ saucy main restricted
deb-src http://ch.archive.ubuntu.com/ubuntu/ saucy main restricted

But I still get the same error when running: sudo apt-get build-dep python-scipy

I have now read the revised installation instructions for scipy that you mention:

https://github.com/rgommers/scipy/blob/5560680db43ca7f89502eb71657fe747d530337a/INSTALL.rst.txt

But it does not describe any of this, not even the necessary sudo apt-get. It also doesn't mention virtualenv as you say I should use.

To write adequate installation instructions I suggest you use a clean linux installation and then write down every single step you need to take in order to install scipy, and briefly explain why the step is necessary (like the TensorFlow installation instructions). And then make a complete rewrite of the scipy installation instructions instead of just adding to an old one; e.g. the instructions for building scipy from source mentions linux distro bugs that are 10 years old.

**Comment by ev-br** [comment]
> I've tried what you suggest but I get the following error:

> ~ $ sudo apt-get build-dep python-scipy
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: You must put some 'source' URIs in your sources.list

AFAIU, this shouldn't happen and I think it means that your system is somehow broken (this may or may not be related to the use of `sudo pip`, I don't know --- at least I definitely messed up some Ubuntu boxes by using it). Fixing a borked up linux install is I think out of scope of this ticket. Once you have a working system, you might find better answers if you ask on the scipy-dev mailing list. 
Good luck!

**Comment by maniteja123** [comment]
Hi , as @ev-br already mentioned this definitely seems to be problem with your apt-get. Just my 2 cents, did you run ``sudo apt-get update`` after adding the new repos to the sources.list file ?

**Comment by Hvass-Labs** [comment]
Thanks for the suggestions. Unfortunately I still get the same error.

It is possible that I broke the Linux Mint installation somehow when I tried installing and building scipy. But I've already used too much time on this and I cannot justify spending more time on this at the moment.

I would again like to encourage you guys to either simplify the build procedure for scipy, or at least document it better. Although scipy has been in development for 15 years (!) it is still only an alpha version. Perhaps it would progress faster if it was easier to contribute.

Cheers!
